CONTRIBUTING.md

```diff
@@ -0,0 +1,231 @@
+
+    ```bash
+    ./ci/build_COMPONENT.sh <HOST_COMPILER> <CXX_STANDARD> <GPU_ARCHS>
+    ./ci/test_COMPONENT.sh  <HOST_COMPILER> <CXX_STANDARD> <GPU_ARCHS>
+    ```
+
+    For more details on building and testing, refer to the [Building and Testing](#building-and-testing) section below.
+
+4. **Commit Your Changes**:
+
+    Group your changes and commit them with a meaningful commit message:
+
+    ```bash
+    git commit -m "Add a brief description of the change"
+    ```
+### Developer Guides
+
+As CCCL evolves, we are in the process of unifying and consolidating our practices and guidelines across all components. However, due to the unique history and characteristics of each component, there are still aspects specific to each one. To aid your contributions and ensure you're aligned with each component's best practices, we offer detailed developer guides:
+
+#### CUB
+- [CUB Developer Guide](cub/docs/developer_overview.rst) - General overview of the design of CUB internals
+- [CUB Test Overview](cub/docs/test_overview.rst) - Overview of how to write CUB unit tets
+- [CUB Tuning Infrastructure](cub/docs/tuning.rst) - Overview of CUB's performance tuning infrastructure
+- [CUB Benchmarks](cub/docs/benchmarking.rst) - Overview of CUB's performance benchmarks
+
+#### Thrust
+
+Coming soon!
+
+#### libcudacxx
+
+Coming soon!
+
+
+
+These guides delve deeper into the specifics of each library and will be invaluable if you're looking to make substantial changes or additions.
+
+## Building and Testing
+
+CCCL components are header-only libraries. This means there isn't a traditional build process for the library itself. However, before submitting your contributions, it's crucial to build and test your changes against the suite of tests. Dedicated build and test scripts for each component are provided in the `ci/` directory to facilitate this.
+
+### Building
+
+To build the tests, you can use the respective build scripts. Building tests does not require a GPU.
+
+Each script expects the following arguments:
+
+- **HOST_COMPILER**: The host compiler you wish to use (e.g., `g++`, `clang++`).
+- **CXX_STANDARD**: The C++ standard version (e.g., `11`, `14`, `17`, `20`).
+- **GPU_ARCHS**: A semicolon-separated list of CUDA GPU architectures (e.g., `70;85;90`). This uses the same syntax as CMake's [CUDA_ARCHITECTURES](https://cmake.org/cmake/help/latest/prop_tgt/CUDA_ARCHITECTURES.html#prop_tgt:CUDA_ARCHITECTURES):
+   - `70` - both PTX and SASS
+   - `70-real` - SASS only
+   - `70-virtual` - PTX only
+
+- **CUB**:
+    ```bash
+    ./ci/build_cub.sh <HOST_COMPILER> <CXX_STANDARD> <GPU_ARCHS>
+    ```
+
+- **libcudacxx**:
+    ```bash
+    ./ci/build_libcudacxx.sh <HOST_COMPILER> <CXX_STANDARD> <GPU_ARCHS>
+    ```
+
+- **Thrust**:
+    ```bash
+    ./ci/build_thrust.sh <HOST_COMPILER> <CXX_STANDARD> <GPU_ARCHS>
+    ```
+
+    Example:
+    ```bash
+    ./ci/build_cub.sh g++ 14 "70;75;80-virtual"
+    ```
+
```

Comment:
suggestion: the test scripts always build the whole component. for rapid development it might be better to only build individual tests
```suggestion
Note that these scripts always build *all* test for a given component, which might take a significant amount of time. For local development it is often beneficial to focus on a single test at a time. For CUB and Thrust you can build individual tests via:
    ```bash
    cmake --build build/latest --target thrust.cpp.cuda.cpp17.test.adjacent_difference -v
    ```
    You can determine the name of the target from the cmake output of the build script invocation

    For libcu++ we utilize lit so you have to go into the build folder and then run lit manually
    ```bash
    cd build
    lit latest/libcudacxx/test/PATH_TO_THE_TEST_FOLDER -sv --param=std=c++11 -Dcompute_archs="86" -Dexecutor="NoopExecutor()"
    ```
```

Reply:
I'm reluctant to go into this much detail at this point, especially with libcu++ and needing to explain `lit`. I'd like to defer that to a more tailored libcu++ developer guide.

Reply:
Fine by me